File: Java/lib/src/test/java/leetcode/TwoSum/TwoSumTest.java
package leetcode.TwoSum;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class TwoSumTest {
  protected Solution solution = new Solution();

  static Stream<Arguments> provideTestCases() {
    return Stream.of(
        Arguments.of(new int[] {2, 7, 11, 15}, 9, new int[] {0, 1}, "increasing order"),
        Arguments.of(new int[] {3, 2, 4}, 6, new int[] {1, 2}, "different order"),
        Arguments.of(new int[] {3, 3}, 6, new int[] {0, 1}, "same numbers"),
        Arguments.of(new int[] {}, 5, new int[] {}, "empty array"),
        Arguments.of(new int[] {1, 2, 3}, 10, new int[] {}, "no solution"));
  }

  @ParameterizedTest(name = "{2}")
  @MethodSource("provideTestCases")
  void testTwoSum(int[] nums, int target, int[] expected, String displayName) {
    int[] result = solution.twoSum(nums, target);
    assertArrayEquals(expected, result);
  }
}

File: Go/two_sum/two_sum_test.go
package two_sum

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTwoSum(t *testing.T) {
	cases := []struct {
		name   string
		nums   []int
		target int
		want   []int
	}{
		{
			name:   "different numbers",
			nums:   []int{2, 7, 11, 15},
			target: 9,
			want:   []int{0, 1},
		},
		{
			name:   "different order",
			nums:   []int{3, 2, 4},
			target: 6,
			want:   []int{1, 2},
		},
		{
			name:   "same numbers",
			nums:   []int{3, 3},
			target: 6,
			want:   []int{0, 1},
		},
		{
			name:   "no solution",
			nums:   []int{1, 2, 3},
			target: 10,
			want:   []int{},
		},
		{
			name:   "empty slice",
			nums:   []int{},
			target: 5,
			want:   []int{},
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			actual := twoSum(tt.nums, tt.target)
			assert.Equal(t, tt.want, actual, "Test case %s failed", tt.name)
		})
	}
}
